### ISSUE TITLE
Please deprecate (and/or rename) `add_development_dependency`

### ISSUE THREAD

**Issue by e2** [rubygems/rubygems#1104]
The main reasons:
1. It's used incorrectly more often than not
2. It's not even practical for development nowadays (unless `--development-all` is used)
3. It's superseded by Bundler's groups
4. There's no distinction between "dev deps", "test deps", "build deps" and "deploy deps" - and these distinctions are getting more and more important every day (especially "test deps")
5. Development workstations don't need the "minimalism" provided by `gem install --dev`
6. Too easy to accidentally add/remove a dependency from the wrong section (runtime vs dev)
7. Putting development-related gems in a gemspec doesn't make any sense from a dev ops point of view
8. Average development needs are constantly growing and there's already too much stuff in the gemspec
9. Bundler supports easy platform switching during development - while gemspecs are only really useful for managing releases of builds
10. Organizing dependencies between gemspec and Bundler's Gemfile is simply too confusing - and deprecating this method would really help guide people in the right direction

Summary: gem development and testing is not the responsibility of RubyGems - and should never be, because otherwise it's just wasting precious developer hours.

**Comment by chulkilee** [comment]
+1 to mark `add_development_dependency` deprecated and remove in the next major release (v2)

Agree that it might be useful in pre-bundler era. However things changed and any ruby projects should use bundler - including gem development. I don't know what is the merit of having dev dependency in gem spec itself and in gem source.

If it's extraneous feature, then it would be better to deprecate and then remove it to focus on other key features.

> With respect to time invested, maintaining the feature is something I will continue to have to do until the feature can be removed.

I appreciate contributors' efforts on this feature, but I don't think we need to consider such sunken cost. Of course we have to get more feedback and then take phaseout process when decided.

> There are many gems that are written and released without using Bundler in any way. Forcing users to edit two files to build their gems does not follow the DRY principle regardless of how you wish to define dependencies. You still force users to put one type of gem metadata in one place and another type in a separate place.

I get your point but I'd argue that using a gem and developing a gem is different use case. In bundler era all ruby developers are expected to use bundler in their environment. dev dependencies *can be* in such environment by Gemfile. It *doesn't need* to be in gemspec.

> There are many different types of users of RubyGems, not just "dev ops" people. I can't make a decision based on the ideals of one group of users unless it is absolutely overwhelming.

I think all **users* use just runtime dependency. dev dependency is for developer of each gem.


I think the main point is how seriously we consider bundler as a "standard tool"...

**Comment by evanphx** [comment]
I don't see any reason to deprecate it. Having development dependencies is still super useful and there is really no reason to remove them.

**Comment by e2** [comment]
@evanphx - how so? I've always believed that, but now I can't find a single "good" reason where it isn't using the "wrong tool for a given job".

The point I 100% agree with is that given the "age" and scale of RubyGems, deprecation is simply not worth the actual costs.

**Comment by e2** [comment]
#### TL;DR - I agree deprecation is costly, so I think a "new" tool (alongside Bundler and RubyGems) would be a better place for "new features" and "deprecations".

@chulkilee - thanks, you totally understand me!

> I don't know what is the merit of having dev dependency in gem spec itself and in gem source.

And without Bundler, there's no guarantee you're actually using the right versions of build tools to build the gem anyway (while `rake install` can use Bundler implicitly).

> I appreciate contributors' efforts on this feature, but I don't think we need to consider such sunken cost. Of course we have to get more feedback and then take phaseout process when decided.

I understand that just keeping RubyGems working as it is requires a ton of effort few people are even capable of appreciating. Also, there are at least thousands of people "relying" on the feature (even for strange purposes), so any "feedback" will be enormous in quantity (i.e. tons of "complaints").

I think a better option would be a "new" tool that would integrate Bundler and RubyGems functionality in an "integrated" and "recommended" way. Package managers are a tough problem - especially when it comes to backward compatibilty. While it would be confusion to have another tool (given there's Bundler and RubyGems already), but it does give millions of people a choice (backward compatibility vs new features).

Also, a lot of gems out there are "ancient" - adding a deprecation to them would only cause maintainers and users to waste focus.

> I think all *users* use just runtime dependency. dev dependency is for developer of each gem.

I can't agree more.

Gems are for "use", while *repositories* are for development. (In a perfect world, we'd all be using the master branches of everything in production).

Not to mention - almost every development dependency can be optional (contributing/developing does not necessarily imply you have to be able to build a gem yourself on your own workstation). And so if dev/build deps are optional - why should they be expected to be part of a "released specification" (gemspec)?

> I think the main point is how seriously we consider bundler as a "standard tool"...

I'd disagree that it's even about bundler - you can have a Rakefile so that e.g. `rake install:deps` uses `gem install -v <version> <gem>` and other system commands to make sure the dev environment is properly set up (including checking out other repositories, submodules, toolchains, OS packages, IDE plugins, etc).

A gemspec is just too restrictive for dev deps - you can't lock to a specific version of Golang for example (e.g. `zeus` gem requires golang properly set up to build it) and you can't use checked out repositories. Even plain Rake is is more flexible and also more "standard" for this.

**Comment by evanphx** [comment]
@e2 What features are you thinking a new tool needs? We've been adding new things people need to Rubygems fine.

Because dev_deps is unable to articulate the universe of development needs doesn't mean it's a bad option. It's not uncommon when working on a gem for the first time to install the development dependencies and the best way they're articulated is via dev_deps.

**Comment by e2** [comment]
@evanphx - a new tool would "absorb" paradigm changes faster than the heavily relied on RubyGems with a massive legacy or expensive (backward-compatibility) features.

As for dev_deps in a gemspec - I believe the fact that no extremely knowledgeable and experienced Ruby "guru" (e.g. you or @drbrain) can articulate a single "good" use case for it is ... case in point.

It's not a bad option technically (I'd just say it's plain inferior given other options) - it's just "cargo cult". It's like a configuration option that no one uses, but everyone specifies because "other people do it".

Rake and Bundler are much better for setting up the dev environment - whether those deps are in the gemspec or not is just an implementation detail. There's just no advantage to having those in a gemspec - only disadvantages.

This is especially true for platform-specific dev tools (Rubinius, JRuby, etc.).

It's pretty much the same story as with specifying test files in a gemspec. People do it but ... what's the point?

> It's not uncommon when working on a gem for the first time to install the development dependencies and the best way they're articulated is via dev_deps.

Not true - the best way to install deps is by what's recommended in the docs. E.g. even Bundler (which can't use itself to install deps) has custom code (for `rake spec:deps`):

https://github.com/bundler/bundler/blob/f2c712b6f5a16ea8f81b432537d2852bbea2df67/Rakefile#L53-L75

And because it uses the gemspec (a mistake in my opinion), it breaks the DRY principle (that @drbain talked about), because it has to remove `ronn` and `rdiscount` on JRuby. This is a clear example of why the gemspec is a *bad* place to store gem deps - except maybe for gems that don't have documentation to generate, etc..

**Comment by e2** [comment]
@evanphx - 

> What features are you thinking a new tool needs? 

An important feature would be "weak dependencies", especially "conflicts", meaning that gem X at version Y causes problems if used - but is not necessary as a dependency (I don't remember if I searched the issues for this, but I'm sure it's popped up). This would allow to have "optional" gems, but also lock to correct versions if they are actually installed.

Another would be "features" - so that gems could depend on a "feature" of a gem, and not on a whole gem - "features" could have conflicts, etc. (you could install multiple "conflicting" gems, but that would be fine if the needed "features" are not in conflict). The alternative would be to have one gem per feature - but that would duplicate a lot of code (breaking DRY), be a maintenance nightmare (lots of gems to track, even though most of the functionality overlaps) and make it hard to "discover" related/available gems.

But I'm discouraged from even suggesting ideas like that if there's so much debate and trouble (though justified) deprecating a single existing method.

**Comment by evanphx** [comment]
@e2 Bundler was designed to solve a very real and specific problem that Rubygems did not: namely a single place to specify all gems for an app that was resolved as an atomic unit.

As for the other features you're discussing, they're interesting and have been brought up over the years. Either they've got unsolved ramifications (if gem A requests feature B that 2 gems provide, which one is activated?) or haven't been explained to provide any tangible value (optional features).

Perhaps I'm not explaining the case for dev_deps enough if you claim that I can't come up with a use. But even if I'm not explaining it well enough, I feel it still provides more value than an optional gems feature.

I hope this doesn't come across as hostile, that's not my intention. We're open to changes within Rubygems but it's important the changes be well considered.

**Comment by e2** [comment]
#### TL;DR - "optional runtime dep info in gemspec" = bad *but* "dev deps in gemspec" = good? I don't get it.

@evanphx -

> Bundler was designed to solve a very real and specific problem that Rubygems did not: namely a single place to specify all gems for an app that was resolved as an atomic unit.

Yes - and this was so useful in practice, some of this functionality is being merged back into RubyGems (reading a Gemfile, etc.), while Bundler outgrew it's purpose (e.g. you can quickly setup a new gem project using Bundler's features). It's now more of a development tool than RubyGems will ever be - and thus my point. 

> As for the other features you're discussing, they're interesting and have been brought up over the years. 

I'm sure they've been brought up - the point I was making, is that I don't believe any of those feature would get implemented in the next 3-5 years anyway (which is a long time in the IT industry) - no matter how well specified they get. And at that point, the paradigm will be so different again, that there may be a "new" different tool (like Bundler) which will superseed the needed functionality.

> Perhaps I'm not explaining the case for dev_deps enough if you claim that I can't come up with a use.  

I understood the example - I just don't see how that's superior e.g. to having dev deps in a Gemfile and using Bundler. Especially since "development" practically means you have the source tree checked out from a repo anyway. I mean - if everyone suddenly moved their dev deps from gemspecs to Gemfiles - I don't see how the world would break or how that would negatively impact productivity.

> But even if I'm not explaining it well enough, I feel it still provides more value than an optional gems feature.

Optional gems are useful if you have a plugin architecture - and there are dependencies between plugins and/or if the plugins can be used as standalone gems. The only other option is to pull in all plugins (makes no sense if they are unrelated) or users will get runtime errors (due to missing conditional deps) - or - if you use complex sets of "microgems" and "metagems" (which the user has to more carefully select in their project = more work). Not to mention that you'd get a circular dependency (plugin depends on app, app depends on plugin - to maintain semver on both sides).

But the value in storing effectively a development configuration in a "product specification"? And where the dependencies are used only for installation and not activation? I don't see the benefit.

Since Ruby is an interpreter (and not a compiler), activation is more important than any "listed" installation dependencies (as you pointed out with the example about activation). And dev deps in a gemspec are *not* about activation (unless there's a Gemfile, etc.).

I don't get the difference: somehow adding development dependencies (not even used when building a gem) in a gem's metadata is more useful than putting optional *runtime* dependencies in a gem (which would allow a lib to do checks at runtime).

Having "optional runtime deps" would be extremely useful, since an outdated gem could work out that it's outdated (in the given context) - even with no RubyGems support for the actual checking.

Why would it be ok to hardcode runtime version checking inside a lib (for supporting optional features/gems), but somehow putting development dep checking in the gem's metadata is "useful"?

Having it both ways is double standards to me.

> I hope this doesn't come across as hostile, that's not my intention. 

No, I apologize if I seem hostile or "trolling". It's baffling to me. To over-exaggerate: it's like a conspiracy - "everyone" acts like it's a well-thought-out super-useful feature, but no one can explain why (or I'm too dumb to understand). 

Putting extra runtime-related metadata in a gemspec seems "evil",  except for development deps, which seem "too sacred to even touch".

It's like Franz Kafka's "The Trial". 

Given everyone's time, I feel guilty for even starting the topic.

I expected one of the following resolutions:
1. "yes, it's useless legacy, but it's too much effort to deprecate"
2. "yes, it's useless legacy, but due to release cycles this won't be deprecated soon"
3. "yes, it's useless legacy - PRs would be appreciated"
4. "It's a very important feature which boosts productivity more than any alternative by doing X, it's used by 99% of the users downloading the gem, without it, the world will break, etc."
5. "We need this or the "I-hate-Bundler" people will start sending us death threats"

I'm closing this issue, because I believe it will sooner become irrelevant (by how technology advances) than coming to a resolution.

**Comment by dgutov** [comment]
> To over-exaggerate: it's like a conspiracy - "everyone" acts like it's a well-thought-out super-useful feature, but no one can explain why (or I'm too dumb to understand).

+1

I see zero upside for defining development dependencies inside gemspec. It's by all measurements easier to define them inside Gemfile, and the result is more flexible, easier to change, etc. So, if we're invoking DRY, that's where the functionality should live.

**Comment by rlue** [comment]
@e2 @drbrain @evanphx 

I published my very first gem recently, and in my search for guidance, stumbled upon this thread. I learned a lot from reading through it, and I thought you might be interested to know that it formed the basis for [a blog post I just wrote about effective communication](http://ryanlue.com/posts/2017-03-01-not-getting-stonewalled). 

I don't point any fingers or take any sides, and I hope you guys find it valuable.

**Comment by ioquatix** [comment]
I was trying to figure this out recently too.

I've always used `add_development_dependency` for things I believe are required for running tests, building the gem, but not for actually using the gem.

I'm not sure if that still makes sense today. It would be great if there was some guidance on this.
